controller_manager/include/controller_manager/control_loop.hpp

```diff
@@ -0,0 +1,54 @@
+// Copyright 2020 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef CONTROLLER_MANAGER__CONTROL_LOOP_HPP_
+#define CONTROLLER_MANAGER__CONTROL_LOOP_HPP_
+
+#include <chrono>
+
+#include "rclcpp/duration.hpp"
+
+/**
+ * \brief      Run a control loop at a set period
+ * \param[in]  period The period of the loop in nanoseconds
+ * \param[in]  now invocable that returns rclcpp::Time
+ * \param[in]  ok invocable that returns true if the loop should continue
+ * \param[in]  do_work invocable that takes the current time and measured period
+ *             and does the work of the control loop
+ * \param[in]  sleep_until invocable that takes a rclcpp::Time and sleeps until
+ *             that time
+ */
+template <
+  typename NowInvocable, typename OkInvocable, typename WorkInvocable, typename SleepInvocable>
+void ControlLoop(
```

Comment:
It would not be testable if put in the `.cpp` file because I need to create a new version of it in the translation unit that is the test. This is the reason that templates always have to go in a header file if you want to use them in two separate translation units.

Reply:
I added static_assert statements to this to make the compile errors nicer if a user tries to use this function and makes a mistake in the sort of function-like objects they pass into this.

Reply:
> Furthermore, do we really need methods as arguments? IMHO, for this simple functionality this looks very overengineered.

Do you have a suggestion for how to make this behavior testable without methods as arguments? IMHO, code that is not designed to be tested is almost always broken.

Reply:
> code that is not designed to be tested is almost always broken.

bit heavy-handed, isn't it?

I agree testability is good, but you cannot generalise like you're doing here. It isn't very constructive.

Reply:
I find it is much harder to make mistakes myself when I write tests. I find writing correct code without tests is often more difficult than writing code with tests. I find that writing code that can be tested, while always more work than the untestable variant, is less work for others to use... As it is less likely someone will need to correct my mistakes later.

Obviously writing tests doesn't eliminate all mistakes, it just makes them more difficult to create.

Statements about tests here may contain hyperbole.

Reply:
There is plenty of working code without tests. If the maintainers of this project would rather this be fixed without tests with a smaller PR I can trivially provide that PR. Much of the value of these tests is that they helped me reason about how to write this and what the corner cases of it is.

Reply:
I'm sorry for being sarcastic and mean here. Please let me know what feels overly complex or over engineered. I'm happy to make changes or add documentation to make this code more readable or better in any way.

Reply:
There is no need to apologise Tyler.

Obviously introducing tests is better than not doing it, and adapting code structure to allow for testing is better than not doing it.

There are different ways of testing though, and unit tests are only one of them. Not all of them require dependency injection or mocking.

Reply:
I tried testing this loop externally with an integration test and was unable to figure out how to do it robustly or simply. My current robot driver I'm working on is effectively a test of this but I do think it is way too much code for a test.

Reply:
You're misunderstanding me: I'm not telling you this is not a good approach, or you should use a different testing strategy.

I'm saying asserting that only "testable/tested code is fault-free" is not a claim you can make, and that using a different approach than dependency injection/mocking could also be used in some cases.

I'm not saying in *this* case.

Reply:
"testable/tested code is fault-free" is an exaggeration for sure. As a person who makes a lot of mistakes, and finds writing code that does what I want hard, I find writing tests useful.

Theoretically, tests also help you in the future when you come back to something and want to make a change but not break what already exists.